### ISSUE TITLE
Switch to std::shared_ptr

### ISSUE THREAD

**Issue by raovgarimella** [amanzi/amanzi#68]
Given that Amanzi has switched to C++11, maybe its time to switch from Teuchos::RCP to shared_ptr?

**Comment by lipnikov** [comment]
I support this suggestion

**Comment by jd-moulton** [comment]
I asked Ross Bartlett if there was a reason to (or not to) switch over.   Here's his reponse:

_The Teuchos Memory Management classes have an optional debug-mode that does a lot of safety checking (and catching and reporting circular references) that std::shared_ptr does not have.  And the standard library has no equivalent class to Teuchos::Ptr in helping to eliminate undefined behavior from C++ programs.  Therefore, a debug-mode build of a code that uses the Teuchos Memory Management classes will run much faster and catches more errors than a program that uses std::shared_ptr and raw pointers run with valgrind.
 
Also, the Teuchos Memory Management classes are about to be made thread-safe by default with pretty similar performance to std::shared_ptr for common usage patterns.  Once that is done, there will be no functional advantage of std::shared_ptr over Teuchos::RCP.
 
Therefore, given the Teuchos Memory Management classes are interoperable with std::shared_ptr, I would not recommend that people switch over (unless they want to drop a required dependence on Trilinos and Teuchos)._

So since we can't eliminate our dependency on Trilinos all together, seems like we would not gain anything, but potentially lose some debugging capabilities.

**Comment by raovgarimella** [comment]
I can't recall an instance where we debugged using the extra
capabilities of Teuchos::RCP classes but I cannot state confidently that
no one in the Amanzi/ATS world hasn't.

**Comment by lipnikov** [comment]
Agree with Rao. Focus on debugging is important but probably not for Amanzi with its more robust hierarchy of classes.
Elimination of RCP is a graduate processes anyway.

**Comment by ecoon** [comment]
This is old, but I would argue that it is more important to remove unnecessary and sometimes misleading uses of RCP first.  We (me especially?) tended to default to passing things by RCP that shouldn't have been.  RCPs should only be passed when we want to make it specifically possible for the accepting thing to take ownership of the same data as the currently owning thing.  As in, really, share ownership.  We almost never actually mean that.  I would guess that 25% of our use of RCP should be replaced with unique_ptr (which eliminates the need for managing destructors like RCP but isn't shared) and another 50% could be replaced by either const or non-const reference or Ptr (in cases where we don't really expect or want the acceptor to keep our object). 

Finally, every time an RCP actually IS passed, it should be passed by const reference to avoid repeatedly incrementing/decrementing the counter, doing more checks, etc.  There are a few places where these are currently passed by value, which is bad.

**Comment by ecoon** [comment]
Specifically, see https://www.youtube.com/watch?v=xnqTKD8uD64 starting at 7 min, or specific to this, jump to 12:30.

**Comment by ecoon** [comment]
I would actually propose an edit to this proposal, and then would like to take a more serious effort at doing some of this.  I would argue for the following guidelines on smart pointers and argument passing:

Unless you can show that it is a computational burden, do not use raw new/delete.  Instead use std::make_unique or Teuchos::rcp.

All factories should return std::unique_ptr<Obj> or Teuchos::RCP<Obj>.  Prefer unique_ptr (you can always make an RCP from a unique_ptr), but use RCP if it will always immediately get stored as an RCP (i.e. State).

Calling signatures:

unique_ptr<X> factory();              // creates an X
void sink( unique_ptr<X>& );       // consumes X, and the caller cannot keep X or use after call
void reseat( unique_ptr<X>& );   // change which X the caller points to (we almost never use this?)

Teuchos::RCP<X> factory();       // creates an X
Teuchos::RCP<X> share();          // the caller will or might keep X
void share( const Teuchos::RCP<X>& );   // the callee will or might keep X
void reseat( Teuchos::RCP<X>& );            // change which X the caller points to 
                                                                     // (we almost never use this)

void optional( X* );                      // if X is an optional argument, i.e. pointer can be nullptr.

void else_if_primitive(X);            // if none of the above apply, and X is a primitive 
void else_if_not_primitive( X& ); // if none of the above apply, and X is not a primitive
X = primitive_return();                // if primitive
X& = non_primitive_return();     // if there is no ownership transfer implied, the caller just 
                                                    // looks at (const) or modifies (nonconst) X but does not keep it.

Note that X can be replaced by const X with no loss of generality.

Note that the key part of this is that passing by RCP should mean something about ownership, i.e. that the callee might keep X!  This also suggests to NEVER use Teuchos::Ptr<X>, instead using X*.  Using Ptr requires that the object WAS stored in an RCP, and means an object stored in a unique_ptr must jump through hoops to be put in a non-owning Ptr before it can be passed.

This is effectively a summary of this: 
https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/

with RCP replacing shared_ptr, though I am perfectly happy to lose RCP in favor of shared_ptr.

**Comment by lipnikov** [comment]
I'd like to use the Newton solver for small problems of size 3-10. Wrapping small non-persistent vectors in RCP poiters looks like an overkill. In view of Ethan's comments, it implies member function  Solve(Vector& u) instead of Solve(const Teuchos::RCP<Vector>& u). This triggers a long chain of changes in most PKs, so I'd like first to collect opinions on this topic.

**Comment by ecoon** [comment]
Yes, I think Solve(Vector& u) is right, and yes, it breaks a lot of code.  

Is your issue with Solve() or with ResidualFunction() or with both?  If it's just with Solve(), then it really isn't an issue because we could just overload, construct a non-owning rcpFromRef(), and call Solve() with that RCP.  If it is ResidualFunction(), UpdatePreconditioner(), and the whole BDFFnBase interface, then it's a much bigger set of code to change, including all PKs, tons of tests, etc etc.

**Comment by lipnikov** [comment]
I think, it is both. I am at the stage of writing a unit test related to the generalized dual porosity model and both Solve() and Residual() are connected. It is not urgent now and I can survive with the RCP pointers. But if there is an agreement that Solve(Vector&) is the way to go then I can plan for a PR after the coming release is complete.
